pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,492 @@
+			}
+		}
+		consistentHashMap := newNodeConsistentHashMap()
+		consistentHashMap.Add(aliveAndMatchedNodes...)
+		c.consistentHashRWMutex.Lock()
+		defer c.consistentHashRWMutex.Unlock()
+		c.consistentHashMap[eip.Name] = consistentHashMap
+		c.notify(eip.Name)
+		return nil
+	}
+
+	if err := updateConsistentHash(eip); err != nil {
+		klog.ErrorS(err, "Sync consistentHash by externalIPPool failed", "externalIPPool", eip.Name)
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+func (c *Cluster) handleClusterNodeEvents(nodeEvent *memberlist.NodeEvent) {
+	node, event := nodeEvent.Node, nodeEvent.Event
+	switch event {
+	case memberlist.NodeJoin, memberlist.NodeLeave:
+		// when a Node joins cluster, all matched externalIPPools consistentHash should be updated.
+		// when a Node leave cluster, the Node may failed or deleted,
+		// if Node has been deleted, affected externalIPPool should enqueue, deleteNode handler has processed
+		// if the Node failed, externalIPPools consistentHash maybe changed, affected externalIPPool should enqueue
+		coreNode, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// means Node has been deleted, deleteNode handler has processed
+				klog.ErrorS(err, "Processing clusterNode events", "eventType", event)
+				return
+			}
+			klog.ErrorS(err, "Processing clusterNode events get node failed", "eventType", event)
+			return
+		}
+		affectedEIPs := c.filterEIPsFromNodeLabels(coreNode)
+		c.enqueueExternalIPPools(affectedEIPs)
+		klog.InfoS("Processed Node event", "eventType", event, "nodeName", node.Name, "affectedExternalIPPoolNum", len(affectedEIPs))
+	case memberlist.NodeUpdate:
+		klog.InfoS("Node event: update node", "nodeName", node.Name)
+	default:
+		klog.InfoS("Node event: unknown", "nodeName", node.Name, "event", event)
+	}
+}
+
+// aliveNodes return alive nodeName list in cluster
+func (c *Cluster) aliveNodes() sets.String {
+	aliveMembers := c.mList.Members()
+	nodes := make([]string, len(aliveMembers))
+	for i, node := range aliveMembers {
+		nodes[i] = node.Name
+	}
+	return sets.NewString(nodes...)
+}
+
+// ShouldSelectEgress only Node which joined cluster need to hold egress
+func (c *Cluster) ShouldSelectEgress(egress *v1alpha2.Egress) bool {
+	eipName := egress.Spec.ExternalIPPool
+	if eipName == "" {
+		return false
+	}
+	// local Node must matches externalIPPool
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		klog.ErrorS(err, "Select node for egress failed", "externalIPPool", eipName)
+		return false
+	}
+	localNode, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		klog.ErrorS(err, "Select node for egress failed", "nodeName", c.nodeName)
+		return false
+	}
+	if !c.matchNodeSelector(localNode, eip) {
+		klog.ErrorS(err, "Select node for egress eip not match", "nodeName", c.nodeName)
```

Comment:
Why using error here? Isn't normal that a node is not selected by a pool?

Reply:
done